### PR TITLE
fix(vite-node): tolerate error in watch mode, fix #1715

### LINKED ISSUE #1715: [vite-node] Do not crash on error in watch mode
### Clear and concise description of the problem

If you run vite-node in watch mode (e.g. `vite-node -w app.ts`), any time your code crashes (e.g. saved with invalid syntax) you have to restart `vite-node` manually.

### Suggested solution

If an error (invalid syntax, code exception etc.) is encountered, **do not stop** the `vite-node` process and wait for the next file change to rerun (potentially fixed) code.

### Alternative

Running `vite-node -w app.ts` manually every time you accidentally saved a file in an invalid state.

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by andrew-w-ross**
This pr is a partial fix for watch mode in `vite-node`, It allows the user to recover on a bad transform. It however won't attempt to keep a process running if it fails think forever or pm2.

### CODE REVIEW

File: packages/vite-node/src/client.ts [hunk 1]
```diff
@@ -129,7 +129,13 @@ export class ViteNodeRunner {
...
   async executeFile(file: string) {
-    return await this.cachedRequest(`/@fs/${slash(resolve(file))}`, [])
+    try {
+      return await this.cachedRequest(`/@fs/${slash(resolve(file))}`, [])
+    }
+    catch (err) {
+      if (this.options.exitOnFailedTransform)
+        throw err
+    }
   }
```
Comment (antfu) [inline packages/vite-node/src/client.ts:139]: I think we should do the error handling on the `cli.ts` caller side instead

Comment (andrew-w-ross) [reply packages/vite-node/src/client.ts]: @antfu Only getting back to this pr now, let me know if this works for you.

File: packages/vite-node/src/cli.ts [hunk 2]
```diff
@@ -75,8 +75,15 @@ async function run(files: string[], options: CliOptions = {}) {
...
-  for (const file of files)
-    await runner.executeFile(file)
+  for (const file of files) {
+    try {
+      await runner.executeFile(file)
+    }
+    catch (e) {
+      if (!options.watch)
+        throw e
+    }
```
Comment (antfu) [inline packages/vite-node/src/cli.ts:85]: I think we should still print out the error and set exitcode

**Comment by antfu** [comment]
And I think we should have an option to make this opt-in.

File: packages/vite-node/src/cli.ts (hunk 2 above)
Comment (andrew-w-ross) [reply packages/vite-node/src/cli.ts:92]: Sure any need to set the exit code back to 0 on hmr?

File: packages/vite-node/src/cli.ts [hunk 3]
```diff
@@ -75,8 +75,22 @@ async function run(files: string[], options: CliOptions = {}) {
...
+  for (const file of files) {
+    try {
+      await runner.executeFile(file)
+    }
+    catch (e) {
+      if (options.watch) {
+        process.exitCode = 1
+        // Circular reference throws a TypeError
+        if (e instanceof Error && e.name === 'TypeError')
+          continue
```
Comment (antfu) [inline packages/vite-node/src/cli.ts:87]: I am not sure about this. Can you give more specific examples of why this is necessary?

**Pull request closed without merging**